Foundation/NSExpression.swift

```diff
@@ -28,40 +28,152 @@ public enum NSExpressionType : UInt {
+        _evaluationBlock = evaluation
+        _substitutionVariablesBlock = substitution
+
+        super.init()
+    }
+
+    public convenience init(forConstantValue value: AnyObject?) {
+        self.init(type:.ConstantValueExpressionType, evaluation:{_,_ in return value})
+
+        _constantValue = value
+    } // Expression that returns a constant value
+
+    public class func expressionForEvaluatedObject() -> NSExpression {
+        return NSExpression(type:.EvaluatedObjectExpressionType, evaluation:{o,_ in return o})
+    } // Expression that returns the object being evaluated
+
+    public convenience init(forVariable aVariable: String) {
+        let evaluationBlock:EvaluateBlockType = {obj,bindings in
+            guard let exp = bindings?.objectForKey(aVariable.bridge()) as? NSExpression else {
+                preconditionFailure("Cannot find variable \(aVariable) in \(bindings)")
+            }
+
+            return exp.expressionValueWithObject(obj, context: bindings)
+        }
+
+        let substitutionVariablesBlock:SubstituteBlockType = {bindings in
+            guard let exp = bindings.objectForKey(aVariable.bridge()) as? NSExpression else {
+                preconditionFailure("Cannot find variable in \(bindings)")
+            }
+
+            return exp
+        }
+
+        self.init(type:.VariableExpressionType, evaluation:evaluationBlock, substitution:substitutionVariablesBlock)
+
+        self._variable = aVariable
+    } // Expression that pulls a value from the variable bindings dictionary
+
+    public convenience init(forKeyPath keyPath: String) {
+        let evaluationBlock:EvaluateBlockType = {obj,_ in
+            // TODO: Change NSObjectProtocol to whatever protocol -valueForKeyPath() belongs to.
+            guard let o = obj where o is NSObjectProtocol else {
+                return nil
+            }
+
+            // TODO: depends on NSObject -valueForKeyPath implementation
+            //return o.valueForKeyPath(keyPath)
+            NSUnimplemented()
+        }
+
+        self.init(type:.KeyPathExpressionType, evaluation:evaluationBlock)
+
+        _keyPath = keyPath
+
+    } // Expression that invokes valueForKeyPath with keyPath
+
+    public convenience init?(forFunction name: String, arguments parameters: [NSExpression]) {
+
+        guard let fn = _NSExpressionFunctions[name] else {
+            // TODO: failable init. Decide if we should throw instead. (Apple Foundation throws an Exception)
+            return nil
+        }
+
+        let evaluationBlock:EvaluateBlockType = {obj, bindings in
+            var result:AnyObject?
+            let args = parameters.expressionValueWithObject(obj, context: bindings)
+
+            do {
+                 result = try _invokeFunction(fn, with:args)
+            } catch let NSExpressionError.InvalidArgumentType(expected: type){
+                // TODO: Decide if we should throw instead. (Apple Foundation throws an Exception)
```

Comment:
throw is reserved for error cases, exception cases are fatalError or precondition